First draft of android readme (#42)
godot-rust/gdext-book: src/toolchain/export-android.md

<!--
  ~ Copyright (c) godot-rust; Bromeon and contributors.
  ~ This Source Code Form is subject to the terms of the Mozilla Public
  ~ License, v. 2.0. If a copy of the MPL was not distributed with this
  ~ file, You can obtain one at https://mozilla.org/MPL/2.0/.
-->

> NOTE: Keep track of where you install the Android Command Line Tools and JDK 17

Reviewer: Several things:
- Please start a page with a h1 heading (Markdown: `# Title`) and add it to SUMMARY.md. 
- The h1 should be followed by an intro section, which is followed by a TOC (table of contents) comment.
- Note blocks should be inside "admonish" tags with a title.

You can check how other pages do it 🙂

Reviewer: The `NOTE` here still needs an admonish block, see previous message :slightly_smiling_face:

> Exporting with gdext for godot requires some of the same pieces that are required for building godot from source.

Reviewer: Capitalization: "Godot" is always capitalized (whole page).

Exporting with gdext for Godot requires some of the same pieces that are required for building Godot from source.
Specifically, the Android SDK Command Line Tools and JDK 17 as mentioned in Godot's documentation
[here](https://docs.godotengine.org/en/stable/contributing/development/compiling/compiling_for_android.html#requirements)

Once you have those installed, you then need to follow Godot's instructions for setting up the build system

> [here](https://docs.godotengine.org/en/stable/contributing/development/compiling/compiling_for_android.html#setting-up-the-buildsystem)

Reviewer: Both should end with .

I have not tested without it.  To be safe, it is included in the documentation for now.

> Now on to the fun part of actually compiling gdext for Android! :)

Reviewer: Avoid "I" in the tutorial, the book has many authors/contributors.

I assume you used JDK 17 because it's mentioned in the Godot docs?

More neutrally, the above could be written as:

> It's not entirely clear if JDK 17 is required for Android Command Line Tools to function properly; it might be possible without it.

Please also avoid emojis/emoticons and sentences like "Now on to the fun part of actually compiling gdext for Android! :)". Instead, this could be a h2 header.

> Set the environment variable CLANG_PATH to point to Android's build of clang.  Example:

Reviewer: Use single backticks for symbols like `CLANG_PATH`.

Avoid double spaces (before "Example") -- you have several occurrences of it in other places, too.

> CLANG_PATH is used by bindgen's clang-sys dependency.  Documentation here: https://github.com/KyleMayes/clang-sys?tab=readme-ov-file#environment-variables 
> 
> ```bash
> #!/bin/bash
> 
> export CLANG_PATH = $"[android-cli-directory]/[android-cli-version]/ndk/[ndk-version]/toolchains/llvm/prebuilt/[host-machine-os]/bin/clang"
> ```

Reviewer: Use links in text when possible. Use curly braces and camelCase for placeholders.

Also write some text that puts the bash code block into context. The `#!/bin/bash` is not needed since it's not a full script, just a snippet.
```suggestion
`CLANG_PATH` is used by bindgen's clang-sys dependency. See also [clang-sys documentation][clang-sys-docs].
<some text introducing the bash section>:

```bash
export CLANG_PATH = $"{androidCliDirectory}/{androidCliVersion}/ndk/{ndkVersion}/toolchains/llvm/prebuilt/{hostMachineOs}/bin/clang"
```

Reviewer: And at the end of the page or section:
```
[clang-sys-docs]: https://github.com/KyleMayes/clang-sys?tab=readme-ov-file#environment-variables 
```

export CLANG_PATH = "{androidCliDirectory}/{androidCliVersion}/ndk/{ndkVersion}/toolchains/llvm/prebuilt/{hostMachineOs}/bin/clang"

> ```

Reviewer: Could you split this with `\` after the `=`?
 At its current length, it likely needs a horizontal scrollbar.

> Then set the ``CARGO_TARGET_[target-triple-here]_LINKER`` to point to the android linker for the android triple you are targetting.  You need to compile your gdext library for each android triple individually.  You can find possible targets by running:

Reviewer: - Capitalization of "Android"
- No double spaces
- Single backticks
- `{placeholderCamelCase}`
- "targeting", not "targetting"

```
rustup target list

> ```

Reviewer: Use ` ```bash ` here

> You can find the linkers in the android cli directory at:

Reviewer: ```suggestion
You can find the linkers in the Android CLI directory at:
```

```

You can find the linkers in the Android CLI directory at:

```text
"{androidCliDirectory}/{androidCliVersion}/ndk/{ndkVersion}/toolchains/llvm/prebuilt/{hostMachineOs}/bin/{targetTripleHere}{androidVersion}"

> ```

Reviewer: Also this could be split (without `\`, at one of the `/` points). 

And the quotes are not needed.

`targetTripleHere` could be just `targetTriple`.

> ```

Reviewer: - ` ```text` 
- no leading `$`
- placeholder as mentioned above

> As of writing this, the tested tuples are:

Reviewer: You mean triples, no?

> | Tuple                       | Environment Variable                            | Godot Arch     | Gdextension Config            |

Reviewer: ```suggestion
| Tuple                       | Environment Variable                            | Godot Arch     | GDExtension Config            |
```

Also use single backticks everywhere. Please apply this to the rest of the page, too.

| Tuple                       | Environment Variable                            | Godot Arch     | GDExtension Config            |
| --------------------------- | ----------------------------------------------- | -------------- | ----------------------------- |
| `aarch64-linux-android`     | `CARGO_TARGET_AARCH64_LINUX_ANDROID_LINKER`     | `arm64`        | `android.debug.arm64`         |
| `x86_64-linux-android`      | `CARGO_TARGET_X86_64_LINUX_ANDROID_LINKER`      | `x86_64`       | `android.debug.x86_64`        |
| `armv7-linux-androideabi`   | `CARGO_TARGET_ARMV7_LINUX_ANDROID_LINKER`       | `arm32`        | `android.debug.armeabi-v7a`   |
| `i686-linux-android`        | `CARGO_TARGET_I686_LINUX_ANDROID_LINKER`        | `x86_32`       | `android.debug.x86`           |

> Notice how the environment variables are in all caps and the triple's "-" is replaced with "_"

Reviewer: ```suggestion
Notice how the environment variables are in all-caps and the triple's "-" is replaced with "_".
```

> Make sure to add all of the triples you want to support to the ``rustup`` via:

Reviewer: ```suggestion
Make sure to add all of the triples you want to support to `rustup` via:
```

```bash
rustup target add {androidTripleHere}

> ```

Reviewer: Make sure you name variables consistently throughout the page, otherwise it's confusing.

This one was referenced as `{targetTriple}`, no? (as mentioned, omit the "here" part).

Example:

```bash
rustup target add aarch64-linux-android
```

> Putting it all together, here is an example compiling for `aarch64-linux-android`. This is also probably the most common
> Android target as of the writing of this.

Reviewer: Could benefit from a new section. Also small correction:

```suggestion
## A complete example

Putting it all together, here is an example compiling for `aarch64-linux-android`. This is also probably the most common
Android target, as of writing of this.
```

> 1. android cli is installed in my HOME folder

Reviewer: ```suggestion
1. Android CLI is installed in the `$HOME` folder.
```

Assuming 2 things:

1. Android CLI is installed in the `$HOME` folder.

> 2. Godot is still relying on Android ndk version: 23.2.8568313. Check

Reviewer: ```suggestion
2. Godot is still relying on Android NDK version 23.2.8568313. Check
```

[here](https://github.com/godotengine/godot/blob/4a0160241fd0c1e874e297f6b08676cf0761e5e8/platform/android/java/app/config.gradle#L14).
3. The downloaded Android CLI version is: 11076708_latest (update this to be the version you downloaded).

> 4. This is being ran on linux. Change the `linux-x86_64` folder in `CLANG_PATH` and `CARGO_TARGET_AARCH64_LINUX_ANDROID_LINKER`

Reviewer: ```suggestion
4. This is being run on Linux. Change the `linux-x86_64` folder in `CLANG_PATH` and `CARGO_TARGET_AARCH64_LINUX_ANDROID_LINKER`
```

to be your host machine's operating system.

> 5. You are targeting Android version 34.

Reviewer: Why "2 things" and then 5 points?

Also, if you really want to use numbers, use low ones like 2 as text "two". But I think "multiple", "following", etc. is typically better, because it doesn't break when someone adds/removes a bullet point.

> 5. You are targetting android version 34

Reviewer: ```suggestion
5. You are targeting Android version 34.
```

> rustup target add aarch64-linux-android

Reviewer: Introduce this code block with surrounding text. If it's not a script, omit `#!/bin/bash`.

And here is what the commands look like running from a bash shell:

```bash
rustup target add aarch64-linux-android

export CLANG_PATH="$HOME/android-cli/11076708_latest/ndk/23.2.8568313/toolchains/llvm/prebuilt/linux-x86_64/bin/clang"
export CARGO_TARGET_AARCH64_LINUX_ANDROID_LINKER=\
    "$HOME/android-cli/11076708_latest/ndk/23.2.8568313/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android34-clang"

cargo build --target=aarch64-linux-android
```

> And then you should find a built version of your GDExtension lib in:

Reviewer: ```suggestion
And then you should find a built version of your GDExtension library in:
```

<!--
 ~ Copyright (c) godot-rust; Bromeon and contributors.
 ~ This Source Code Form is subject to the terms of the Mozilla Public
 ~ License, v. 2.0. If a copy of the MPL was not distributed with this
 ~ file, You can obtain one at https://mozilla.org/MPL/2.0/.
-->

# Export to Android

```admonish note title="Android CLI and JDK 17"
Keep track of where you install the Android Command Line Tools and JDK 17

> ```

Reviewer: What does "keep track" mean? Notes should ideally be pieces of knowledge, or actionable.

Author: Ya I guess this part really isn't needed.  Its common sense that you need to know where the software you want to run is installed.

Exporting with gdext for Godot requires some of the same pieces that are required for building Godot from source.
Specifically, the Android SDK Command Line Tools and JDK 17 as mentioned in Godot's documentation
[here](https://docs.godotengine.org/en/stable/contributing/development/compiling/compiling_for_android.html#requirements).

Once you have those installed, you then need to follow Godot's instructions for setting up the build system
[here](https://docs.godotengine.org/en/stable/contributing/development/compiling/compiling_for_android.html#setting-up-the-buildsystem).

As of writing this, it is not entirely clear if JDK 17 is required in order for Android Command Line Tools to function properly.

> It has not been not tested without it. To be safe, it is included in the documentation for now.

Reviewer: Is there a way to clarify this? Google doesn't contain installation instructions?

I'd rather avoid users needing to install extra toolchains or higher versions of Java if those turn out to be unnecessary.

Author: It looks like there stance is:

> Make sure to choose a JDK version that is higher than or equal to the JDK versions used by plugins that you use in your Gradle build. To determine the minimum required JDK version for the Android Gradle Plugin (AGP), see the compatibility table in the [release notes](https://developer.android.com/build/releases/gradle-plugin).
>
> For example, the Android Gradle Plugin version 8.x requires JDK 17. If you try to run a Gradle build that uses it with an earlier version of the JDK, it reports a message like:

https://developer.android.com/build/jdks

So I guess we have to reference what godot is using.  Lets see if we can find it.

Found it!

https://github.com/godotengine/godot/blob/master/platform/android/java/app/config.gradle

Looks like they have godot configured for java version 17.  I'll update the docs to reference the version godot is using.

Reviewer: Very nice, thanks a lot!

## Compiling

The environment variable `CLANG_PATH` is used by bindgen's clang-sys dependency. See also
[clang-sys documentation](https://github.com/KyleMayes/clang-sys?tab=readme-ov-file#environment-variables)

Set the environment variable `CLANG_PATH` to point to Android's build of clang. Example:

```bash
export CLANG_PATH=\
"{androidCliDirectory}/{androidCliVersion}/ndk/{ndkVersion}/toolchains/llvm/prebuilt/{hostMachineOs}/bin/clang"
```

Then set the `CARGO_TARGET_{targetTriple}_LINKER` to point to the Android linker for the Android triple you are targeting.

> You need to compile your gdext library for each Android triple individually. Possible targets can be found by running:

Reviewer: Actually, it seems like `{targetTriple}` is formatted differently than below, so we should use a different name.

Maybe `{shoutTargetTriple}` and explain that it's `SHOUT_CASE`. You could adjust the existing "notice how the environment..." sentence accordingly.

| `aarch64-linux-android`     | `CARGO_TARGET_AARCH64_LINUX_ANDROID_LINKER`     | `arm64`        | `android.debug.arm64`         |
| `x86_64-linux-android`      | `CARGO_TARGET_X86_64_LINUX_ANDROID_LINKER`      | `x86_64`       | `android.debug.x86_64`        |
| `armv7-linux-androideabi`   | `CARGO_TARGET_ARMV7_LINUX_ANDROID_LINKER`       | `arm32`        | `android.debug.armeabi-v7a`   |
| `i686-linux-android`        | `CARGO_TARGET_I686_LINUX_ANDROID_LINKER`        | `x86_32`       | `android.debug.x86`           |

Notice how the environment variables are in all-caps and the triple's "-" is replaced with "_".

Make sure to add all of the triples you want to support to `rustup` via:

```bash
rustup target add {targetTriple}
```

Example:

```bash
rustup target add aarch64-linux-android
```

## A complete example

Putting it all together, here is an example compiling for `aarch64-linux-android`. This is also probably the most common
Android target, as of the writing of this.

Assuming the following things:

1. Android CLI is installed in the `$HOME` folder.
2. Godot is still relying on Android NDK version 23.2.8568313. Check

> [here](https://github.com/godotengine/godot/blob/4a0160241fd0c1e874e297f6b08676cf0761e5e8/platform/android/java/app/config.gradle#L14).

Reviewer: Should this maybe be a link to `master` rather than a commit?

Then the user can verify if it's still correct.

Author: Yep!  I can do that the same way I found out how to find the java version to use

Author: done :)

3. The downloaded Android CLI version is: 11076708_latest (update this to be the version you downloaded).
4. This is being run on Linux. Change the `linux-x86_64` folder in `CLANG_PATH` and `CARGO_TARGET_AARCH64_LINUX_ANDROID_LINKER`
to be your host machine's operating system.
5. You are targeting Android version 34.

And here is what the commands look like running from a bash shell:

```bash
rustup target add aarch64-linux-android

export CLANG_PATH="$HOME/android-cli/11076708_latest/ndk/23.2.8568313/toolchains/llvm/prebuilt/linux-x86_64/bin/clang"
export CARGO_TARGET_AARCH64_LINUX_ANDROID_LINKER=\
"$HOME/android-cli/11076708_latest/ndk/23.2.8568313/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android34-clang"

cargo build --target=aarch64-linux-android
```

And then you should find a built version of your GDExtension library in:

```text
target/aarch64-linux-android/debug/{libraryName}.so

> ```

Reviewer: I just commented in the macOS/iOS PR that we should probably use `{YourCrate}` as the universal placeholder to refer to the user's crate, so it's consistent across pages.

Author: will do :)

Author: done :)
